README.md
[![Build Status](https://travis-ci.org/pedrohff/flutter_heatmap_calendar.svg?branch=master)](https://travis-ci.org/pedrohff/flutter_heatmap_calendar)
# Flutter Heat Map Calendar
A Heat Map Calendar based on Github's contributions chart which can be used to visualize values over time

![HeatMap Calendar in action](https://thumbs.gfycat.com/ImmaculateRequiredLarva.webp)


## Installing

### 1. Depend on it
Add this to your package's pubspec.yaml file:

```
dependencies:
  heatmap_calendar: ^1.2.8
```

### 2. Install it
You can install packages from the command line:

with pub:

```shell
$ pub get
```

with Flutter:

```shell
$ flutter pub get
```

## Example
```dart
import 'package:heatmap_calendar/heatmap_calendar.dart';
import 'package:heatmap_calendar/time_utils.dart';
...

HeatMapCalendar(
  input: {
    TimeUtils.removeTime(DateTime.now().subtract(Duration(days: 3))): 5,
    TimeUtils.removeTime(DateTime.now().subtract(Duration(days: 2))): 35,
    TimeUtils.removeTime(DateTime.now().subtract(Duration(days: 1))): 14,
    TimeUtils.removeTime(DateTime.now()): 5,
  },
  colorThresholds: {
    1: Colors.green[100],
    10: Colors.green[300],
    30: Colors.green[500]
  },
  weekDaysLabels: ['S', 'M', 'T', 'W', 'T', 'F', 'S'],
  monthsLabels: [
    "",
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
  ],
  squareSize: 16.0,
  textOpacity: 0.3,
  labelTextColor: Colors.blueGrey,
  dayTextColor: Colors.blue[500],
)
```


CHANGELOG.md
## [1.2.8] - 2020-07-23

* Fixing date difference for timezones

## [1.2.7] - 2020-07-01

* Removing generated files

## [1.2.6] - 2020-05-19

* Support of unfinished week
* Add more tests

## [1.2.5] - 2019-07-06

* Fixing issues when creating the calendar columns (validating if datesList is not empty)

## [1.2.4] - 2019-05-31

* Fixing issues

## [1.2.3] - 2019-05-31

* Fixing tests

## [1.2.2] - 2019-05-31

* Fixing width problems

## [1.2.1] - 2019-05-31

* Fixing unused import

## [1.2.0] - 2019-05-31

* Creating tests for all classes
* Separated WeekColumns logic to a new Widget

## [1.1.2] - 2019-05-27

* Fixing margin on week labels

## [1.1.1] - 2019-05-27

* Fixing the amount of columns that are being created

## [1.1.0] - 2019-05-26

* Building columns based on the layout available widget

## [1.0.2] - 2019-05-25

* Adding Travis CI integration

## [1.0.1] - 2019-05-25

* Formatting code and fixing pub requirements

## [1.0.0] - 2019-05-25

* Release of the package

test/time_utils_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:heatmap_calendar/time_utils.dart';

final DateTime startWeek = DateTime(2020, 5, 31);
final DateTime friday = DateTime(2020, 6, 5);

void main() {
  group('Testing the firstDayOfTheWeek method', () {
    test('should return weekday as 7', () {
      expect(TimeUtils.firstDayOfTheWeek(friday).weekday, equals(7));
    });

    test('should return the first day of the week', () {
      expect(
          TimeUtils.firstDayOfTheWeek(friday),
          equals(startWeek));
    });

    test('should return same day if its first day of the week', () {
      expect(
          TimeUtils.firstDayOfTheWeek(startWeek),
          equals(startWeek));
    });
  });

  group('Testing the removeTime method', () {
    test('dateTime should be at 00:00:00.000', () {
      DateTime dateTime = TimeUtils.removeTime(DateTime.now());
      expect(dateTime.hour, equals(0));
      expect(dateTime.minute, equals(0));
      expect(dateTime.second, equals(0));
      expect(dateTime.millisecond, equals(0));
      expect(dateTime.microsecond, equals(0));
    });
  });

  group('Testing the dateBetween method', () {
    test('list should not be null', () {
      List<DateTime> datesBetween = TimeUtils.datesBetween(
          DateTime.now().subtract(Duration(days: 7)), DateTime.now());

      expect(datesBetween, isNotNull);
    });

    test('list should not be empty', () {
      List<DateTime> datesBetween = TimeUtils.datesBetween(
          DateTime.now().subtract(Duration(days: 7)), DateTime.now());

      expect(datesBetween, isNotEmpty);
    });

    test('should return a list with given date amount', () {
      List<DateTime> datesBetween = TimeUtils.datesBetween(
          DateTime.now().subtract(Duration(days: 7)), DateTime.now());

      expect(datesBetween.length, equals(8));
    });

    test('should have the initial value', () {
      DateTime initialDate = DateTime.now().subtract(Duration(days: 7));
      List<DateTime> datesBetween =
          TimeUtils.datesBetween(initialDate, DateTime.now());

      expect(datesBetween, contains(initialDate));
    });

    test('should fail assertion if initial date is after finish date', () {
      expect(
          () => TimeUtils.datesBetween(
              DateTime.now(), DateTime.now().subtract(Duration(days: 7))),
          throwsAssertionError);
    });

    test('No duplicated dates in DST change', () {
      List<DateTime> datesBetween = TimeUtils.datesBetween(DateTime(2020, 10, 23), DateTime(2020, 10, 27));
      Set<int> set = datesBetween.map((date) => date.day).toSet();
      expect(set.length, datesBetween.length);
    });

    test('All dates set to midnight in DST change', () {
      List<DateTime> datesBetween = TimeUtils.datesBetween(DateTime(2020, 10, 23), DateTime(2020, 10, 27));
      datesBetween.forEach((date) => expect(date.hour, 0, reason: date.toIso8601String()));

      List<DateTime> datesBetweenSummer = TimeUtils.datesBetween(DateTime(2020, 3, 27), DateTime(2020, 3, 31));
      datesBetweenSummer.forEach((date) => expect(date.hour, 0, reason: date.toIso8601String()));
    });
  });
}


test/heatmap_day_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:heatmap_calendar/heatmap_day.dart';

void main() {
  group('HeatMapDay Widget', () {
    Widget subject;
    Widget app;
    Color defaultColor = Colors.green;
    Map<int, Color> thresholds = {
      1: Colors.red[100],
      30: Colors.red[300],
      60: Colors.red[700]
    };

    setUp(() {
      subject = HeatMapDay(
        value: 10,
        size: 5.0,
        thresholds: thresholds,
        currentDay: DateTime.now().day,
        fontWeight: FontWeight.normal,
        defaultColor: defaultColor,
      );

      app = new MaterialApp(home: subject);
    });

    testWidgets('should have only one HeatMapDay', (WidgetTester tester) async {
      await tester.pumpWidget(app);

      expect(find.byType(HeatMapDay), findsOneWidget);
    });

    testWidgets('should have only one Container', (WidgetTester tester) async {
      await tester.pumpWidget(app);

      expect(find.byType(Container), findsOneWidget);
    });

    testWidgets('should have only one AnimatedOpacity',
        (WidgetTester tester) async {
      await tester.pumpWidget(app);

      expect(find.byType(AnimatedOpacity), findsOneWidget);
    });

    testWidgets('should have only one Text', (WidgetTester tester) async {
      await tester.pumpWidget(app);

      expect(find.byType(Text), findsOneWidget);
    });

    testWidgets('should find one widget with given text',
        (WidgetTester tester) async {
      await tester.pumpWidget(app);

      expect(find.text(DateTime.now().day.toString()), findsOneWidget);
    });
  });

  group('Testing method getColorFromThreshold', () {
    Color defaultColor = Colors.green;
    Map<int, Color> thresholds = {
      1: Colors.red[100],
      30: Colors.red[300],
      60: Colors.red[700]
    };

    HeatMapDay subject = HeatMapDay(
      value: 10,
      size: 5.0,
      thresholds: thresholds,
      currentDay: DateTime.now().day,
      fontWeight: FontWeight.normal,
      defaultColor: defaultColor,
    );

    test('should return default color if no value is given', () {
      subject = HeatMapDay(
        size: 5.0,
        thresholds: thresholds,
        currentDay: DateTime.now().day,
        fontWeight: FontWeight.normal,
        defaultColor: defaultColor,
      );
      Color color = subject.getColorFromThreshold();
      expect(color, equals(defaultColor));
    });

    test('should return the default color if value is zero', () {
      subject = HeatMapDay(
        value: 0,
        size: 5.0,
        thresholds: thresholds,
        currentDay: DateTime.now().day,
        fontWeight: FontWeight.normal,
        defaultColor: defaultColor,
      );
      Color color = subject.getColorFromThreshold();
      expect(color, equals(defaultColor));
    });

    test('should return the first color', () {
      subject = HeatMapDay(
        value: 5,
        size: 5.0,
        thresholds: thresholds,
        currentDay: DateTime.now().day,
        fontWeight: FontWeight.normal,
        defaultColor: defaultColor,
      );
      Color color = subject.getColorFromThreshold();
      expect(color, equals(Colors.red[100]));
    });

    test('should return the last color', () {
      subject = HeatMapDay(
        value: 100,
        size: 5.0,
        thresholds: thresholds,
        currentDay: DateTime.now().day,
        fontWeight: FontWeight.normal,
        defaultColor: defaultColor,
      );
      Color color = subject.getColorFromThreshold();
      expect(color, equals(Colors.red[700]));
    });
  });
}


test/month_label_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:heatmap_calendar/month_label.dart';

void main() {
  String text = "20";

  MonthLabel subject = MonthLabel(
    size: 5.0,
    textColor: Colors.black,
    text: text,
  );

  Widget app = MaterialApp(
    home: Scaffold(
      body: subject,
    ),
  );

  group('Testing the MonthLabel widget', () {
    testWidgets('should have only one MonthLabel', (tester) async {
      await tester.pumpWidget(app);

      expect(find.byType(MonthLabel), findsOneWidget);
    });

    testWidgets('should have only one Container', (tester) async {
      await tester.pumpWidget(app);

      expect(find.byType(Container), findsOneWidget);
    });

    testWidgets('should have only one Stack inside the Container',
        (tester) async {
      await tester.pumpWidget(app);

      expect(
          find.descendant(
              of: find.byType(Container), matching: find.byType(Stack)),
          findsOneWidget);
    });

    testWidgets('should have only one Positioned', (tester) async {
      await tester.pumpWidget(app);

      expect(find.byType(Positioned), findsOneWidget);
    });

    testWidgets('should have only one Text', (tester) async {
      await tester.pumpWidget(app);

      expect(find.byType(Text), findsOneWidget);
    });

    testWidgets('should have a widget with given text', (tester) async {
      await tester.pumpWidget(app);

      expect(find.text(text), findsOneWidget);
    });
  });
}


lib/time_utils.dart
class TimeUtils {
  /// The first element is an empty string,
  /// once Dart's DateTime counts months from 1 to 12
  static const List<String> defaultMonthsLabels = [
    "",
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
  ];

  static const List<String> defaultWeekLabels = [
    'S',
    'M',
    'T',
    'W',
    'T',
    'F',
    'S'
  ];

  /// Obtains the first day of the current week,
  /// based on user's current day
  static DateTime firstDayOfTheWeek(DateTime today) {
    return safeSubtract(today, Duration(
        days: (today.weekday % DateTime.daysPerWeek),
        hours: today.hour,
        minutes: today.minute,
        seconds: today.second,
        microseconds: today.microsecond,
        milliseconds: today.millisecond));
  }

  static DateTime firstDayOfCalendar(DateTime day, int columnsAmount) {
    return safeSubtract(day, Duration(days: (DateTime.daysPerWeek * (columnsAmount - 1))));
  }

  /// Sets a DateTime hours/minutes/seconds/microseconds/milliseconds to 0
  static DateTime removeTime(DateTime dateTime) {
    return DateTime(dateTime.year, dateTime.month, dateTime.day);
  }

  /// Returns date without timezone info (UTC format)
  static DateTime removeTZ(DateTime dateTime) {
    return DateTime.utc(dateTime.year, dateTime.month, dateTime.day, dateTime.hour, dateTime.minute,
        dateTime.second, dateTime.millisecond, dateTime.millisecond);
  }

  /// Returns date with local timezone
  static DateTime addTZ(DateTime dateTime) {
    return DateTime(dateTime.year, dateTime.month, dateTime.day, dateTime.hour, dateTime.minute,
        dateTime.second, dateTime.millisecond, dateTime.microsecond);
  }

  /// Subtract duration without timezone.
  /// That prevents from problems with time shift if DST changed
  static DateTime safeSubtract(DateTime dateTime, Duration duration) {
    DateTime add = removeTZ(dateTime).subtract(duration);
    return addTZ(add);
  }

  /// Add duration without timezone.
  /// That prevents from problems with time shift if DST changed
  static DateTime safeAdd(DateTime dateTime, Duration duration) {
    DateTime add = removeTZ(dateTime).add(duration);
    return addTZ(add);
  }

  /// Creates a list of [DateTime], including all days between [startDate] and [finishDate]
  static List<DateTime> datesBetween(DateTime startDate, DateTime finishDate) {
    assert(startDate.isBefore(finishDate));

    List<DateTime> datesList = new List();
    DateTime aux = startDate;
    do {
      datesList.add(aux);
      aux = safeAdd(aux, Duration(days: 1));
    } while (finishDate.millisecondsSinceEpoch >= aux.millisecondsSinceEpoch);

    return datesList;
  }
}


lib/heatmap_day.dart
import 'package:flutter/material.dart';

class HeatMapDay extends StatelessWidget {
  final int value;
  final double size;
  final Map<int, Color> thresholds;
  final Color defaultColor;
  final int currentDay;
  final double opacity;
  final Duration animationDuration;
  final Color textColor;
  final FontWeight fontWeight;

  const HeatMapDay(
      {Key key,
      this.value,
      this.size,
      this.thresholds,
      this.defaultColor: Colors.black12,
      this.currentDay,
      this.opacity: 0.3,
      this.animationDuration: const Duration(milliseconds: 300),
      this.textColor: Colors.black,
      this.fontWeight})
      : super(key: key);

  /// Loop for getting the right color based on [thresholds] values
  ///
  /// If the [value] is greater than or equal one of [thresholds]' key,
  /// it will receive its value
  Color getColorFromThreshold() {
    Color color = defaultColor;
    if (value != null) {
      thresholds.forEach((mapKey, mapColor) {
        if (value >= mapKey) {
          color = mapColor;
        }
      });
    }

    return color;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      alignment: Alignment.center,
      height: size,
      width: size,
      color: getColorFromThreshold(),
      margin: EdgeInsets.all(2.0),
      child: AnimatedOpacity(
        opacity: opacity,
        duration: animationDuration,
        child: Text(
          currentDay.toString(),
          style: TextStyle(fontWeight: fontWeight, color: textColor),
        ),
      ),
    );
  }
}


lib/week_labels.dart
import 'package:flutter/material.dart';
import 'package:heatmap_calendar/default_container.dart';

class WeekLabels extends StatelessWidget {
  final List<String> weekDaysLabels;
  final double squareSize;
  final Color labelTextColor;

  const WeekLabels(
      {Key key,
      @required this.weekDaysLabels,
      @required this.squareSize,
      @required this.labelTextColor})
      : assert(weekDaysLabels != null),
        assert(weekDaysLabels.length == 7),
        assert(squareSize != null),
        assert(squareSize > 0),
        assert(labelTextColor != null),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        DefaultContainer(
          text: "",
          size: squareSize,
          textColor: labelTextColor,
          margin: 0,
        ),
        DefaultContainer(
          text: weekDaysLabels[0],
          size: squareSize,
          textColor: labelTextColor,
        ),
        DefaultContainer(
          text: weekDaysLabels[1],
          size: squareSize,
          textColor: labelTextColor,
        ),
        DefaultContainer(
          text: weekDaysLabels[2],
          size: squareSize,
          textColor: labelTextColor,
        ),
        DefaultContainer(
          text: weekDaysLabels[3],
          size: squareSize,
          textColor: labelTextColor,
        ),
        DefaultContainer(
          text: weekDaysLabels[4],
          size: squareSize,
          textColor: labelTextColor,
        ),
        DefaultContainer(
          text: weekDaysLabels[5],
          size: squareSize,
          textColor: labelTextColor,
        ),
        DefaultContainer(
          text: weekDaysLabels[6],
          size: squareSize,
          textColor: labelTextColor,
        ),
      ],
    );
  }
}


lib/week_columns.dart
import 'package:flutter/material.dart';
import 'package:heatmap_calendar/heatmap_day.dart';
import 'package:heatmap_calendar/month_label.dart';
import 'package:heatmap_calendar/time_utils.dart';

class WeekColumns extends StatelessWidget {
  final double squareSize;

  final Color labelTextColor;

  final Map<DateTime, int> input;

  final Map<int, Color> colorThresholds;

  final double currentOpacity;

  final List<String> monthLabels;

  final Color dayTextColor;

  final int columnsToCreate;

  final DateTime date;

  const WeekColumns(
      {Key key,
      @required this.squareSize,
      @required this.labelTextColor,
      @required this.input,
      @required this.colorThresholds,
      @required this.currentOpacity,
      @required this.monthLabels,
      @required this.dayTextColor,
      @required this.columnsToCreate,
      @required this.date})
      : super(key: key);

  /// The main logic for generating a list of columns representing a week
  /// Each column is a week having a [MonthLabel] and 7 [HeatMapDay] widgets
  List<Widget> buildWeekItems() {
    List<DateTime> dateList = getCalendarDates(columnsToCreate);
    int totalDays = dateList.length;
    var daysPerWeek = DateTime.daysPerWeek;
    int totalWeeks = (totalDays / daysPerWeek).ceil();
    int amount = totalDays + totalWeeks;

    // The list of columns that will be returned
    List<Widget> columns = new List();

    // The list of items that will be used to form a week
    List<Widget> columnItems = new List();
    List<int> months = new List();

    for (int i = 0; i < amount; i++) {
      // If true, it means that it should be a label,
      // if false, it should be a HeatMapDay
      if (i % 8 == 0) {
        String month = "";

        if (dateList.isNotEmpty && !months.contains(dateList.first.month)) {
          month = monthLabels[dateList.first.month];
          months.add(dateList.first.month);
        }

        columnItems.add(MonthLabel(
          size: squareSize,
          textColor: labelTextColor,
          text: month,
        ));
      } else {
        DateTime currentDate = dateList.first;
        dateList.removeAt(0);

        final int value = (input[currentDate] == null) ? 0 : input[currentDate];

        HeatMapDay heatMapDay = HeatMapDay(
          value: value,
          thresholds: colorThresholds,
          size: squareSize,
          currentDay: currentDate.day,
          opacity: currentOpacity,
          textColor: dayTextColor,
        );
        columnItems.add(heatMapDay);

        // If the columnsItems has a length of 8, it means it should be ended.
        if (columnItems.length == 8) {
          columns.add(Column(children: columnItems));
          columnItems = new List();
        }
      }
    }

    if (columnItems.isNotEmpty) {
      columns.add(Column(children: columnItems));
    }
    return columns;
  }

  /// Creates a list of all weeks based on given [columnsAmount]
  List<DateTime> getCalendarDates(int columnsAmount) {
    DateTime firstDayOfTheWeek = TimeUtils.firstDayOfTheWeek(date);
    DateTime firstDayOfCalendar = TimeUtils.firstDayOfCalendar(firstDayOfTheWeek, columnsAmount);
    return TimeUtils.datesBetween(firstDayOfCalendar, date);
  }

  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: buildWeekItems(),
      ),
    );
  }
}


lib/heatmap_calendar.dart
import 'package:flutter/material.dart';
import 'package:heatmap_calendar/heatmap_day.dart';
import 'package:heatmap_calendar/time_utils.dart';
import 'package:heatmap_calendar/week_columns.dart';
import 'package:heatmap_calendar/week_labels.dart';

class HeatMapCalendar extends StatefulWidget {
  static const double COLUMN_COUNT = 11;
  static const double ROW_COUNT = 8;
  static const double EDGE_SIZE = 4;

  /// The labels identifying the initials of the days of the week
  /// Defaults to [TimeUtils.defaultWeekLabels]
  final List<String> weekDaysLabels;

  /// The labels identifying the months of a year
  /// Defaults to [TimeUtils.defaultMonthsLabels]
  final List<String> monthsLabels;

  /// The inputs that will fill the calendar with data
  final Map<DateTime, int> input;

  /// The thresholds which will map the given [input] to a color
  ///
  /// Make sure to map starting on number 1, so the user might notice the difference between
  /// a day that had no input and one that had
  /// Example: {1: Colors.green[100], 20: Colors.green[200], 40: Colors.green[300]}
  final Map<int, Color> colorThresholds;

  /// The size of each item of the calendar
  final double squareSize;

  /// The opacity of the text when the user double taps the widget
  final double textOpacity;

  /// The color of the texts in the weeks/months labels
  final Color labelTextColor;

  /// The color of the text that identifies the days
  final Color dayTextColor;

  /// Helps avoiding overspacing issues
  final double safetyMargin;

  const HeatMapCalendar(
      {Key key,
      @required this.input,
      @required this.colorThresholds,
      this.weekDaysLabels: TimeUtils.defaultWeekLabels,
      this.monthsLabels: TimeUtils.defaultMonthsLabels,
      this.squareSize: 16,
      this.textOpacity: 0.2,
      this.labelTextColor: Colors.black,
      this.dayTextColor: Colors.black,
      this.safetyMargin: 0})
      : super(key: key);

  @override
  HeatMapCalendarState createState() {
    return HeatMapCalendarState();
  }
}

class HeatMapCalendarState extends State<HeatMapCalendar> {
  double currentOpacity = 0;
  bool displayDates = false;

  /// Toggles the labels in all [HeatMapDay]s
  void onDoubleTap() {
    setState(() {
      displayDates = !displayDates;
      currentOpacity = displayDates ? widget.textOpacity : 0;
    });
  }

  /// Calculates the right amount of columns to create based on [maxWidth]
  ///
  /// returns the number of columns that the widget should have
  int getColumnsToCreate(double maxWidth) {
    assert(maxWidth > (2 * (HeatMapCalendar.EDGE_SIZE + widget.squareSize)));

    // The given size of a square + the size of the margin
    final double widgetWidth = widget.squareSize + HeatMapCalendar.EDGE_SIZE;
    return (maxWidth - widget.safetyMargin) ~/ widgetWidth;
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        return InkWell(
          onDoubleTap: onDoubleTap,
          child: Container(
            height: (widget.squareSize + HeatMapCalendar.EDGE_SIZE) *
                (HeatMapCalendar.ROW_COUNT + 1),
            width: constraints.maxWidth,
            child: Row(
              children: <Widget>[
                WeekLabels(
                  weekDaysLabels: widget.weekDaysLabels,
                  squareSize: widget.squareSize,
                  labelTextColor: widget.labelTextColor,
                ),
                WeekColumns(
                  squareSize: widget.squareSize,
                  labelTextColor: widget.labelTextColor,
                  input: widget.input,
                  colorThresholds: widget.colorThresholds,
                  currentOpacity: currentOpacity,
                  monthLabels: widget.monthsLabels,
                  dayTextColor: widget.dayTextColor,
                  columnsToCreate: getColumnsToCreate(constraints.maxWidth) - 1,
                  date: DateTime.now(),
                )
              ],
            ),
          ),
        );
      },
    );
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:heatmap_calendar/heatmap_calendar.dart';
import 'package:heatmap_calendar/time_utils.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: SafeArea(
          child: HeatMapCalendar(
            input: {
              TimeUtils.removeTime(DateTime.now().subtract(Duration(days: 3))): 5,
              TimeUtils.removeTime(DateTime.now().subtract(Duration(days: 2))): 35,
              TimeUtils.removeTime(DateTime.now().subtract(Duration(days: 1))): 14,
              TimeUtils.removeTime(DateTime.now()): 5,
            },
            colorThresholds: {
              1: Colors.green[100],
              10: Colors.green[300],
              30: Colors.green[500]
            },
            weekDaysLabels: ['S', 'M', 'T', 'W', 'T', 'F', 'S'],
            monthsLabels: [
              "",
              "Jan",
              "Feb",
              "Mar",
              "Apr",
              "May",
              "Jun",
              "Jul",
              "Aug",
              "Sep",
              "Oct",
              "Nov",
              "Dec",
            ],
            squareSize: 20.0,
            textOpacity: 0.3,
            labelTextColor: Colors.blueGrey,
            dayTextColor: Colors.blue[500],
          ),
        ),
      ),
    );
  }
}
